lib/shared/bucketing-assembly-script/assembly/eventQueue/requestPayloadManager.ts

```diff
@@ -101,33 +108,46 @@ export class RequestPayloadManager {
             }
         }
 
-        // Generate defaulted agreagate user as our events APIs require a user / user_id
+        // Generate defaulted aggregate user as our events APIs require a user / user_id
         const dvcUser = new DVCPopulatedUser(new DVCUser(
             user_id, null, null, null, null, NaN, null, null, null, null
         ))
-        this.addEventsToPendingPayloads(new UserEventsBatchRecord(dvcUser, aggEvents))
+        return new UserEventsBatchRecord(dvcUser, aggEvents)
     }
 
     /**
-     * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
+     * Find existing FlushPayload that is less than this.chunkSize, and isn't a failed payload
      */
-    private addEventsToPendingPayloads(record: UserEventsBatchRecord): void {
-        let flushPayload = new FlushPayload([])
+    private getFlushPayload(): FlushPayload {
+        const payloads = this.pendingPayloads.values()
 
-        while (record.events.length > this.chunkSize) {
-            const batchRecord = new UserEventsBatchRecord(
-                record.user,
-                record.events.splice(0, this.chunkSize)
-            )
-            flushPayload.records.push(batchRecord)
-            this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
-            flushPayload = new FlushPayload([])
-        }
-        if (record.events.length > 0) {
-            flushPayload.records.push(record)
+        for (let i=0; i < payloads.length; i++) {
+            const payload = payloads[i]
+            if (payload.status === 'failed') {
+                continue
+            }
+            if (payload.eventCount() < this.chunkSize) {
+                return payload
+            }
         }
-        if (flushPayload.records.length > 0) {
-            this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
+
+        return new FlushPayload([])
+    }
+
+    /**
+     * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
+     */
+    private addEventRecordsToPendingPayloads(records: UserEventsBatchRecord[]): void {
+        for (let i = 0; i < records.length; i++) {
+            const record: UserEventsBatchRecord = records[i]
+
+            do {
+                const flushPayload = this.getFlushPayload()
+                flushPayload.addBatchRecordForUser(record, this.chunkSize)
+                if (flushPayload.records.length > 0) {
+                    this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
+                }
+            } while (record.events.length > 0)
```

Comment:
I think you still want this to be a regular while loop, there's no reason to run it even once if there are not events in the record